### PR TITLE
[VideoReader] FFMPEG Multithreading, take2

### PR DESCRIPTION
**Pull request by bjuncek**
Addressing issue #3388 on implementing multithreaded video decoding, and superseeding #3389.

### CODE REVIEW

File: torchvision/io/__init__.py [hunk 1]
```diff
@@ -107,7 +111,7 @@ def __init__(self, path, stream="video"):
                 + "ffmpeg (version 4.2 is currently supported) and"
                 + "build torchvision from source."
             )
-        self._c = torch.classes.torchvision.Video(path, stream)
+        self._c = torch.classes.torchvision.Video(path, stream, num_threads)
```
Comment (prabhat00155) [inline torchvision/io/__init__.py:114]: num_threads is not defined anywhere. Should it be an argument in `__init__()`?

File: torchvision/csrc/io/decoder/stream.cpp [hunk 2]
```diff
@@ -49,7 +51,51 @@ int Stream::openCodec(std::vector<DecoderMetadata>* metadata) {
...
+    codecCtx_->thread_count = num_threads;
+
+  } else {
+    // otherwise set sensible defaults
+    // with the special case for the different MPEG4 codecs
+    // that don't have threading context functions
+    // TODO: potentially automate this using native c++ function lookups
+    if (strcmp(codecCtx_->codec->name, "mpeg4") == 0 &&
+        codecCtx_->codec_type == 0) {
+      if (codecCtx_->codec_tag == 1684633208) {
```
Comment (prabhat00155) [inline torchvision/csrc/io/decoder/stream.cpp:77]: What is this magic number here?

File: torchvision/csrc/io/decoder/stream.cpp [hunk 3]
```diff
@@ -49,7 +51,51 @@ int Stream::openCodec(std::vector<DecoderMetadata>* metadata) {
                << ", avcodec_alloc_context3 failed";
     return AVERROR(ENOMEM);
   }
+  // multithreading heuristics
+  // this is a dumb conservative limit; ideally we'd use 
+  // int max_threads = at::get_num_threads(); but this would cause 
+  // fb sync to fail as it would add dependency to ATen to the decoder API
+  int max_threads = 12;
```
Comment (prabhat00155) [inline torchvision/csrc/io/decoder/stream.cpp:58]: Should we do a `#define` for this instead of using a magic number inside a function?

**Comment by prabhat00155** [comment]
Thanks @bjuncek! Could you fix the linter/format errors?

File: torchvision/csrc/io/decoder/stream.cpp (hunk 3 above)
Comment (bjuncek) [reply torchvision/csrc/io/decoder/stream.cpp:58]: I suppose we can yes; is it customary to do that in the header file?

File: torchvision/csrc/io/decoder/stream.cpp (hunk 2 above)
Comment (bjuncek) [reply torchvision/csrc/io/decoder/stream.cpp:77]: This is codec signature for videos from HMDB and some from UCF.
In principle, we can avoid this by checking if the codec implements threading context function, but I'm  not sure how to do it; I'll ask around.

File: torchvision/io/__init__.py (hunk 1 above)
Comment (bjuncek) [reply torchvision/io/__init__.py:114]: Ah yes - forgot to add this

File: torchvision/csrc/io/decoder/stream.cpp (hunk 2 above)
Comment (prabhat00155) [reply torchvision/csrc/io/decoder/stream.cpp:77]: Okay sure, the alternative solution sounds better.

File: torchvision/csrc/io/decoder/stream.cpp (hunk 3 above)
Comment (prabhat00155) [reply torchvision/csrc/io/decoder/stream.cpp:58]: Do you suspect max_threads will be used elsewhere? Perhaps declaring it a `const` might be better(could be under namespace ffmpeg.

**Review by prabhat00155 — COMMENTED** [review]
Thanks @bjuncek! Could you also resolve the merge conflicts and linter errors?

File: torchvision/csrc/io/decoder/stream.cpp (hunk 3 above)
Comment (bjuncek) [reply torchvision/csrc/io/decoder/stream.cpp:58]: No, not really. At least not in the same context.

Comment (bjuncek) [reply torchvision/csrc/io/decoder/stream.cpp:58]: I've defined `max_threads` as the protected const int in the header file until we find a better way of solving this.

File: torchvision/csrc/io/decoder/__init__.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
+import torch
```
Comment (prabhat00155) [inline torchvision/csrc/io/decoder/__init__.py:1]: I thought `csrc` directory is for C++ code, why are you adding `__init__.py` here?

File: torchvision/csrc/io/decoder/__init__.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+        return True
+
+
+else:
+
+    def _has_video_opt() -> bool:
+        return False
+
+
+class VideoReader:
```
Comment (prabhat00155) [inline torchvision/csrc/io/decoder/__init__.py:47]: Isn't this already defined in [`torchvision/io/__init__.py`](https://github.com/pytorch/vision/blob/972ca657416c5896ba6e0f5fe7c0d0f3e99e1087/torchvision/io/__init__.py#L47).

File: torchvision/csrc/io/decoder/video.h [hunk 6]
```diff
@@ -0,0 +1,58 @@
+#pragma once
```
Comment (prabhat00155) [inline torchvision/csrc/io/decoder/video.h:1]: This code looks same as https://github.com/pytorch/vision/blob/main/torchvision/csrc/io/video/video.h.

File: torchvision/csrc/io/decoder/video.cpp [hunk 7]
```diff
@@ -0,0 +1,349 @@
+#include "video.h"
```
Comment (prabhat00155) [inline torchvision/csrc/io/decoder/video.cpp:1]: Is this a copy of https://github.com/pytorch/vision/blob/main/torchvision/csrc/io/video/video.cpp? What are we trying to do here?

**Comment by bjuncek** [comment]
CLosing and sending new one as cleanup

**Pull request closed without merging**